File: Assets/Scripts/Gameplay/MoveFinder.cs
using System.Collections.Generic;
using System.Linq;
using Match3.Data;

namespace Match3.Gameplay
{
    /// <summary>
    /// 扫描棋盘所有合法交换，供 Hint 与 AutoPlayer 使用。
    /// 通过快照模拟，不修改原棋盘状态。
    /// </summary>
    public class MoveFinder
    {
        private readonly Board _board;
        private readonly GameConfig _config;

        public MoveFinder(Board board, GameConfig config)
        {
            _board = board;
            _config = config;
        }

        /// <summary>
        /// 扫描整个棋盘，返回所有合法交换（每对相邻格只出现一次）。
        /// </summary>
        public List<MoveData> FindAllValidMoves()
        {
            var moves = new List<MoveData>();

            for (var x = 0; x < _board.Width; x++)
            {
                for (var y = 0; y < _board.Height; y++)
                {
                    var from = new GridPos(x, y);

                    if (x + 1 < _board.Width)
                        TryAddMove(moves, from, new GridPos(x + 1, y));

                    if (y + 1 < _board.Height)
                        TryAddMove(moves, from, new GridPos(x, y + 1));
                }
            }

            return moves;
        }

        /// <summary>
        /// 返回预估消除格数最多的交换，供 Hint / AutoPlayer 使用。
        /// </summary>
        public MoveData? FindBestMove()
        {
            var moves = FindAllValidMoves();
            if (moves.Count == 0)
                return null;

            return moves.OrderByDescending(m => m.EstimatedMatchCount).First();
        }

        private void TryAddMove(List<MoveData> moves, GridPos from, GridPos to)
        {
            var matchCount = EstimateMatchCount(from, to);
            if (matchCount > 0)
                moves.Add(new MoveData(from, to, matchCount));
        }

        private int EstimateMatchCount(GridPos from, GridPos to)
        {
            var snapshot = _board.CreateSnapshot();
            snapshot.Swap(from, to);

            var matcher = new Matcher(snapshot, _config);
            return matcher.FindAllMatches().Count;
        }
    }
}

File: web/src/core/moveFinder.ts
import { Board } from "./board";
import { Matcher } from "./matcher";
import { canSwap, GameConfig, GridPos, MoveData, gridPos } from "./types";

export class MoveFinder {
  constructor(
    private readonly board: Board,
    private readonly config: GameConfig,
  ) {}

  findAllValidMoves(): MoveData[] {
    const moves: MoveData[] = [];

    for (let x = 0; x < this.board.width; x++) {
      for (let y = 0; y < this.board.height; y++) {
        const from = gridPos(x, y);

        if (x + 1 < this.board.width) {
          this.tryAddMove(moves, from, gridPos(x + 1, y));
        }

        if (y + 1 < this.board.height) {
          this.tryAddMove(moves, from, gridPos(x, y + 1));
        }
      }
    }

    return moves;
  }

  findBestMove(): MoveData | null {
    const moves = this.findAllValidMoves();
    if (moves.length === 0) return null;

    return moves.reduce((best, move) =>
      move.estimatedMatchCount > best.estimatedMatchCount ? move : best,
    );
  }

  private tryAddMove(moves: MoveData[], from: GridPos, to: GridPos): void {
    if (!canSwap(this.board.getPiece(from)) || !canSwap(this.board.getPiece(to))) {
      return;
    }

    const matchCount = this.estimateMatchCount(from, to);
    if (matchCount > 0) {
      moves.push({ from, to, estimatedMatchCount: matchCount });
    }
  }

  private estimateMatchCount(from: GridPos, to: GridPos): number {
    const snapshot = this.board.createSnapshot();
    snapshot.swap(from, to);
    const matcher = new Matcher(snapshot, this.config);
    return matcher.findAllMatches().size;
  }
}
